src/problems/peg_puzzle.clj

```diff
@@ -0,0 +1,188 @@
+      (= 0 (nth board index))
+      (= 1 (nth board (+ (- index row) 1)))
+      (= 1 (nth board (+ (- index (* 2 row)) 3))))))
+
+
+
+(defn legalUL? [board index n]
+  (let [row (rowOfIndex index n)
+        illegal (firstTwoDiagonals n)]
+    (and
+      (= (.indexOf (for [x [index]
+            y illegal]
+          (compare x y)) 0) -1)
+      (= 0 (nth board index))
+      (= 1 (nth board (- index row)))
+      (= 1 (nth board (+ (- index (* 2 row)) 1))))))
+
+
+
+(defn legalLR? [board index n]
+  (let [row (rowOfIndex index n)]
+    (and
+      (not= row n)
+      (not= row (- n 1))
+      (= 0 (nth board index))
+      (= 1 (nth board (+ (+ index row) 1)))
+      (= 1 (nth board (+ (+ index (* 2 row)) 3))))))
+
+
+(defn legalLL? [board index n]
+  (let [row (rowOfIndex index n)]
+    (and
+      (not= row n)
+      (not= row (- n 1))
+      (= 0 (nth board index))
+      (= 1 (nth board (+ index row)))
+      (= 1 (nth board (+ (+ index (* 2 row)) 1))))))
+
+(defn zeros [board]
+  (filter #(= 0 (nth board %))
+        (range (count board))))
+
+(defn jumpR [board index n]
+  (if (legalR? board index n) (swap board index (+ index 1) (+ index 2))))
+
+(defn jumpL [board index n]
+ (if (legalL? board index n) (swap board index (- index 1) (- index 2))))
+
+(defn jumpUL [board index n]
+  (let [row (rowOfIndex index n)]
+   (if (legalUL? board index n) (swap board index (- index row) (+ (- index (* 2 row)) 1)))))
+
+(defn jumpUR [board index n]
+  (let [row (rowOfIndex index n)]
+   (if (legalUR? board index n) (swap board index (+ (- index row) 1) (+ (- index (* 2 row)) 3)))))
+
+(defn jumpLL [board index n]
+  (let [row (rowOfIndex index n)]
+   (if (legalLL? board index n) (swap board index (+ index row) (+ (+ index (* 2 row)) 1)))))
+
+(defn jumpLR [board index n]
+  (let [row (rowOfIndex index n)]
+   (if (legalLR? board index n) (swap board index (+ (+ index row) 1) (+ (+ index (* 2 row)) 3)))))
+
+;start with boardList empty
+;add all jumps to boardList
+;then child runs the entire boardList into states(just have to call child at end)
+(defn children[state]
+  (if (legalBoard? (:board state) (:n state))
+    (let [b (:board state)
+        n (:n state)
+        z (zeros b)
+        rightList (map #(jumpR b % n) z)
+        leftList (map #(jumpL b % n) z)
+        upperRightList (map #(jumpUR b % n) z)
+        upperLeftList (map #(jumpUL b % n) z)
+        lowerRightList (map #(jumpLR b % n) z)
+        lowerLeftList (map #(jumpLL b % n) z)
+        listList (concat rightList leftList upperRightList upperLeftList lowerRightList lowerLeftList)
+        finalList (remove nil? listList)
+        ]
+
+  (map #(->State % n) finalList))))
+
+;this is kind of half of a heuristic because it's only in effect
+;while there are more than half the pegs left in the puzzle...
+(defn heuristic [state]
+  (let [b (:board state)
+        n (:n state)]
+    (if (> (sum b) (quot (reduce + (range n)) 2)) (quot (* n 20) (count (children state))) 0)))
```

Comment:
Fixing the problem with the call to `heuristic-fn` in `a-star` revealed an issue here, where you divide by zero when `(count (children state))` is zero, i.e., `state` is the solution or there are no legal moves from `state`.